common/switchboard.hpp

```diff
@@ -1,142 +1,503 @@
+			}
+		}
+
+		/**
+		 * @brief Publishes @p this_event to the topic
+		 *
+		 * Thread-safe
+		 */
+		void put(ptr<const event>* this_event) {
+			assert(this_event);
+			assert(*this_event);
+
+			/* The pointer that this gets exchanged with needs to get dropped. */
+			ptr<const event>* old_event = _m_latest.exchange(this_event);
+			delete old_event;
+			// corresponds to new in the last iteration of topic::put
+
+			// Read/write on _m_subscriptions.
+			// Must acquire shared state on _m_subscriptions_lock
+			std::unique_lock lock{_m_subscriptions_lock};
+			for (topic_subscription& ts : _m_subscriptions) {
+				ts.enqueue(std::const_pointer_cast<const event>(*this_event));
+			}
+		}
+
+		/**
+		 * @brief Schedules @p callback on the topic (@p plugin_id is for accounting)
+		 *
+		 * Thread-safe
+		 */
+		void schedule(std::size_t plugin_id, std::function<void(ptr<const event>, std::size_t)> callback) {
+			// Write on _m_subscriptions.
+			// Must acquire unique state on _m_subscriptions_lock
+			const std::unique_lock lock{_m_subscriptions_lock};
+			_m_subscriptions.emplace_back(_m_name, plugin_id, callback, _m_record_logger);
+		}
+
+		/**
+		 * @brief Stop and remove all topic_subscription threads.
+		 *
+		 * Thread-safe
+		 */
+		void stop() {
+			// Write on _m_subscriptions.
+			// Must acquire unique state on _m_subscriptions_lock
+			const std::unique_lock lock{_m_subscriptions_lock};
+			_m_subscriptions.clear();
+		}
+
+		~topic() {
+			ptr<const event>* last_event = _m_latest.exchange(nullptr);
+			if (last_event) {
+				delete last_event;
+				// corresponds to new in most recent topic::put
+			}
+		}
+	};
+
+public:
+
+	/**
+	 * @brief A handle which can read the latest event on a topic.
+	 */
+	template <typename specific_event>
+	class reader {
+	private:
+		// Reference to the underlying topic
+		topic& _m_topic;
+
+	public:
+		reader(topic& topic_)
+			: _m_topic{topic_}
+		{
+#ifndef NDEBUG
+			if (typeid(specific_event) != _m_topic.ty()) {
+				std::cerr << "topic '" << _m_topic.name() << "' holds type " << _m_topic.ty().name() << ", but caller used type" << typeid(specific_event).name() << std::endl;
+				abort();
+			}
+#endif
+		}
+
+		/**
+		 * @brief Gets a "read-only" copy of the latest value.
+		 *
+		 * This will return null if no event is on the topic yet.
+		 */
+		 ptr<const specific_event> get_nullable() const {
+			ptr<const event> this_event = _m_topic.get();
+			ptr<const specific_event> this_specific_event = std::dynamic_pointer_cast<const specific_event>(this_event);
+
+			if (this_event) {
+				assert(this_specific_event /* Otherwise, dynamic cast failed; dynamic type information could be wrong*/);
+				return this_specific_event;
+			} else {
+				return ptr<const specific_event>{nullptr};
+			}
+		}
+
+		/**
+		 * @brief Gets a non-null "read-only" copy of the latest value.
+		 *
+		 * @throws If no event is on the topic yet.
+		 */
+		 ptr<const specific_event> get() const {
```

Comment:
What's the intended use case of this? That is, when would a reader know for sure the value is valid?

Reply:
`get()` is intended to return a "checked" pointer. The check-for-null would otherwise be duplicated at every call-site.

Reply:
Maybe the `get` name should mirror the `get_rw` one